src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -629,13 +637,58 @@ trait Namers extends MethodSynthesis {
         )
       }
 
-      sym setInfo {
-        mkTypeCompleter(copyDef) { sym =>
-          assignParamTypes()
-          lazyType complete sym
+      mkTypeCompleter(copyDef) { sym =>
+        assignParamTypes()
+        lazyType complete sym
+      }
+    }
+
+    // for apply/unapply, which may need to disappear when they clash with a user-defined method of matching signature
+    def applyUnapplyMethodCompleter(un_applyDef: DefDef, companionContext: Context): TypeCompleter =
+      new CompleterWrapper(completerOf(un_applyDef)) {
+        override def complete(sym: Symbol): Unit = {
+          super.complete(sym)
+
+          // If there's a same-named locked symbol, we're currently completing its signature.
+          // This means it (may) refer to us, and is thus either overloaded or recursive without a signature.
+          val scopePartiallyCompleted =
+            companionContext.scope.lookupAll(sym.name).exists(existing => existing != sym && existing.hasFlag(LOCKED))
+
+          val suppress =
+            // Check `scopePartiallyCompleted` first to rule out locked symbols from the owner.info.member call,
+            // as FindMember will call info on a locked symbol (while checking type matching to assemble an overloaded type),
+            // and throw a TypeError, so that we are aborted.
+            scopePartiallyCompleted || {
+            val userDefined = companionContext.owner.info.member(sym.name).filter(_ != sym)
+            (userDefined != NoSymbol) && {
+              userDefined.info match {
+                // TODO: do we have something for this already? the synthetic symbol can't be overloaded, right?
+                case OverloadedType(pre, alternatives) =>
+                  // pre probably relevant because of inherited overloads?
+                  alternatives.exists(_.isErroneous) || alternatives.exists(alt => pre.memberInfo(alt) matches pre.memberInfo(sym))
+                case tp =>
+                  (tp eq ErrorType) || tp.matches(sym.info)
+              }
+            }
```

Comment:
As we vaguely discussed, it would be create to have less opaque type completers for method symbols that let us find out the parameter list arity first (which could filter out some members already), then the method type with a wildcard result to represent "not yet inferred", and then finally the full method type. It would still be a bit of a pain to use, as we couldn't just reuse `findMember` as it, we'd need to generalise it to allow us to specify a signature schema to narrow the search.